float32 ref
---
bool success
